mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClusterActor.java

```diff
@@ -209,14 +209,29 @@ public Receive createReceive() {
 
     private void onAddNewJobArtifactsToCacheRequest(AddNewJobArtifactsToCacheRequest req) {
         try {
-            mantisJobStore.addNewJobArtifactsToCache(req.getClusterID(), req.getArtifacts());
-            jobArtifactsToCache.addAll(req.artifacts);
+            List<ArtifactID> newArtifacts = req.artifacts
+                .stream()
+                .filter(artifactID -> !jobArtifactsToCache.contains(artifactID))
+                .collect(Collectors.toList());
+
+            if (!newArtifacts.isEmpty()) {
+                log.info("Storing and caching new artifacts: {}", newArtifacts);
+
+                mantisJobStore.addNewJobArtifactsToCache(req.getClusterID(), newArtifacts);
+                jobArtifactsToCache.addAll(newArtifacts);
+                refreshTaskExecutorJobArtifactCache();
+            }
             sender().tell(Ack.getInstance(), self());
         } catch (IOException e) {
             log.warn("Cannot add new job artifacts {} to cache in cluster: {}", req.getArtifacts(), req.getClusterID(), e);
         }
     }
 
+    private void refreshTaskExecutorJobArtifactCache() {
+        getTaskExecutors(ExecutorStateManager.isAvailable).getTaskExecutors().forEach(taskExecutorID ->
```

Comment:
Should this be the set of registered TEs instead?

Reply:
Good question. From the TE point of view we want  to cache job artifacts only if the TE can still receive a job. If that's not the case we shouldn't since the cache is ephemeral in nature and will get purged at every restart. With that we also make caching very efficient when we add new job artifacts since we only ask available TEs to download at the same time, avoiding a burst of unnecessary calls to s3.  wdyt?

Reply:
We should decouple the TEs implementation details from the control plane's semantics. The current semantics is that we want to ensure all the TEs cache a set of common job artifacts. I think we should make these calls to the TEs based on that. The implementation details of how the TE manages the cache are independent, IMO.